### ISSUE TITLE
Computing sensitivity: how is it done now, and what other options can be implemented?

### ISSUE THREAD

**Issue by Viskari** [PecanProject/pecan#611]
As I've been going through the results for our analysis with the radiation parameters, I have been bothered by a somewhat basic question. When pecan prints out the quantile values, how does it determine those values? The quantile values don't match the file outputs, yet the sensitivity parameter seems to be calculated as the difference between the highest and lowest quantile values in the output.

So for example we are looking at NPP. For late hardwood for a specific parameter, the lowest quantile value as expressed by pecan is 8036.165 units, however when summing together the NPP in the tower analysis file for that quantile, the end value is 15562.91 units. 

I've been going through the pecan code trying to understand this, but thought that should also check here to avoid misunderstanding things.

**Comment by dlebauer** [comment]
What is _supposed_ to happen is for the sensitivity to be computed at the median value. This is done by the function `get.sensitivity` on line 62 of `[modules/uncertainty/R/get.sensitivity`](https://github.com/PecanProject/pecan/blob/master/modules/uncertainty/R/sensitivity.analysis.R#L62)

```r
#--------------------------------------------------------------------------------------------------#
##' Calculate the sensitivity of a function at the median
##'
##' This function evaluates the sensitivity of a model to a parameter.
##' This is done by evaluating the first derivative of the univariate spline estimate
##' of the model response at the parameter median.
##' @name get.sensitivity
##' @title Calculate Sensitivity
##' @param trait.samples 
##' @param sa.splinefun 
##' @export
##' @return numeric estimate of model sensitivity to parameter
get.sensitivity <- function(trait.samples, sa.splinefun){
  sensitivity <- sa.splinefun(median(trait.samples), 1)
}
```

I will go ahead and use this as an opportunity to migrate Redmine issue ... this was to change the sensitivity analysis from computing sensitivity at the mean to computing an average, e.g. 

mean sensitivity =  integral over x p(x) f'(x) dx , or

using `mean(sa.splinefun(trait.samples,1))` instead of `sa.splinefun(median(trait.samples), 1)`

But we should maintain the original functionality (perhaps as default? do we need a new tag in settings?)

(note that the second argument to `sa.splinefun` is the derivative at which the function is evaluated: `sa.splinefun(x)` evaluates the function at `x`; `sa.splinefun(x,1)` evaluates the first derivative at x

Here is a proposal for a new version of the function:
```r
get.sensitivity <- function(trait.samples, sa.splinefun, method = 'median'){
if(method == 'median'){
     sensitivity <- sa.splinefun(median(trait.samples), 1)
} else if (method == 'integrated mean') {
     sensitivity <- mean(sa.splinefun(trait.samples,1))
} else if (method == 'the method suggested here') {
      logger.warn("using", method, " may be hard to justify, see github issue 611")
      range <- quantile(trait.samples, range(settings$senstivity.analysis$quantiles))
      sensitivity <- diff(sa.splinefun(range)) / diff(range)
}
```

**Comment by Viskari** [comment]
The median value was something I was slightly struggling with there, as I
couldn't figure out what the median value was for. Is it the median after
half of the values have been summed together?

**Comment by dlebauer** [comment]
Where does the sum come in?

**Comment by Viskari** [comment]
That is actually my question here, as the sensitivity value can't be the median of the vector values. The maximum value of FMEAN_NPP_PY for this run is 8.96 units. The quantile value for that run pecan prints out is 8036.165 units when looking at the NPP sensitivity. So based on this median cannot be the median value of the vector of values, so what is it the median off? When writing the question I was pondering could the quantile value used in the sensitivity analysis be the integrated NPP up until the median, but the value doesn't match up with that.

So this returns us to the basic question. When pecan prints out the quantile values, what are those quantile values? As they are not values from the vector of values given to pecan.

**Comment by mdietze** [comment]
@Viskari I'm having a hard time following your questions, but the sensitivity analysis in PEcAn is documented in quite a bit of detail in the LeBauer et al Ecological Monographs paper.

The sensitivity analysis is done as a one-at-a-time analysis holding all parameters at their posterior median except the one being varied. The varied parameter is changed to the quantiles you specify. For convenience the system allows you to specify the sigma equivalents to the standard normal (e.g. the quantile equivalents to 1, 2, 3 standard deviations). PEcAn then makes plots of each parameter vs the requested output variable(s).

In the uncertainty analysis (variance decomposition), the single sensitivity metric (which is normalized to an elasticity) is only diagnostic, it's not used to compute the partial variance. This is why I've never worried too much about the two alternatives David's proposed (though I admit that the integrated version is probably better).  This elasticity doesn't involve any sums or differences of the model outputs, it's the first derivative of the spline function fit to those model outputs, evaluated at the median parameter value.

Another thing that would help in general is to report the units you are talking about, rather than saying 8.96 units and 8036.165 units since the units may not be the same in both places.

**Comment by Viskari** [comment]
I understood all of that before, which was why I was having difficulties
understanding the sensitivity analysis results pecan was outputting as it
didn't seem to match with that. However, we think we have figured out what
is causing the issue.

The unit of the FMEAN_NPP_PY vector produced by ED2 is kgC/m2/yr while
pecan output unit is kgC/m2/s. However, when pecan runs the unit conversion
in model2netcdf.ED2.R, it assumes that the ED2 output is per second, not
per year. In addition to that, it also seems to convert from micromoles to
kilograms, even though both outputs are in kilograms to begin with. These
issues with the unit conversion seem to cause the problems I am having with
the sensitivity analysis outputs.

The basic solution to this is for me to just alter the unit conversion in
model2netcdf.ED.R, but first want to check if there are any hesitations
about that? Also I am going to post a question on ED2 site on why they look
at the values at an annual scale as it doesn't seem intuitive to me.

 Another question, this feeding more to David's earlier point, is that
should we move away from using median as the statistic value here as it is
really questionable with deciduous trees? For example, when the leaf
radiation parameters only affect the system when there are leaves, yet for
more than half of the year, there are no leaves in the trees. Thus the
median value would always be from that period of time when there are no
leaves and thus not really tell us really much of interest.

**Comment by mdietze** [comment]
model2netcdf.ED2 should output units in the PEcAn standard (which is at this moment still equivalent to the MsTMIP standard).

The sensitivity is calculated at the MEDIAN PARAMETER VALUE, it is not the median of the outputs. The output being analyzed is the mean output over whatever variable and time window you specify.

**Comment by Viskari** [comment]
My point on model2netcdf.ED2 was that when it runs the unit conversion, the
unit conversion assumes ED2 output variables to be something they are not.
Hence my question was that are there any issues with me just changing how
model2netcdf.ED2 does the unit conversion for git version of ED2?

As for the second part, this still confuses me as my discussion so far has
been on the quantile values, where as I've understood the model is run with
the parameter value at that part of the distribution. Hence I still
struggle how a median is used there to assign the parameter value. However,
I will first see if the fixing the unit conversion solves the issue I am
having before touching on that more.

**Comment by Viskari** [comment]
I was able to figure out why the results I was getting out were so weird and while I was able to address it, it did raise another issue for me.

I was under the impression pecan only did the unit conversion once, in model2netcdf, which converts carbon fluxes to kgC m-2 s-1, but it actually does it twice? With get.results converting the flux values to kgC ha-1 yr-1, while Shawn and I were under the impression that the output was in model2netcdf units.

Why is pecan currently converting the flux terms to those units, as I thought we were using the MsTMIP units?

**Comment by mdietze** [comment]
model2netcdf only does the conversion once, from the model native format to MsTMIP.

The analysis routine converts to kgC ha-1 yr-1 for the purpose of visualization because that's the units that people are much more accustomed to thinking about. Actually, you should double check the units as it might be MgC/ha/yr, which is an even more common unit.

**Comment by Viskari** [comment]
I double checked it, it converts it to kgC/ha/yr.

However, even if we intend to run the conversion, shouldn't we try to do all of the in the same place, especially since the output pdfs, at least the ones I'm getting, don't have the metrics on the y axis. So for example here, after I checked the units from model2netcdf and just assumed those would be what the values would be in.

**Comment by Viskari** [comment]
And as a slight addition, I actually noticed that pecan does actually give NPP units in the ensemble graphs, it just gives them as kgC/m2/s, which as far as I can tell, isn't the correct unit.

**Comment by dlebauer** [comment]
Can this be closed, replaced by #616?
